server-common/src/main/java/org/apache/kafka/server/util/InterBrokerSendThread.java

```
@@ -0,0 +1,253 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import org.apache.kafka.clients.ClientRequest;
+import org.apache.kafka.clients.ClientResponse;
+import org.apache.kafka.clients.KafkaClient;
+import org.apache.kafka.clients.RequestCompletionHandler;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.errors.AuthenticationException;
+import org.apache.kafka.common.errors.DisconnectException;
+import org.apache.kafka.common.internals.FatalExitError;
+import org.apache.kafka.common.utils.Time;
+
+/**
+ * An inter-broker send thread that utilizes a non-blocking network client.
+ */
+public abstract class InterBrokerSendThread extends ShutdownableThread {
+
+    protected volatile KafkaClient networkClient;
+
+    private final int requestTimeoutMs;
+    private final Time time;
+    private final UnsentRequests unsentRequests;
+
+    public InterBrokerSendThread(
+        String name,
+        KafkaClient networkClient,
+        int requestTimeoutMs,
+        Time time
+    ) {
+        this(name, networkClient, requestTimeoutMs, time, true);
+    }
+
+    public InterBrokerSendThread(
+        String name,
+        KafkaClient networkClient,
+        int requestTimeoutMs,
+        Time time,
+        boolean isInterruptible
+    ) {
+        super(name, isInterruptible);
+        this.networkClient = networkClient;
+        this.requestTimeoutMs = requestTimeoutMs;
+        this.time = time;
+        this.unsentRequests = new UnsentRequests();
+    }
+
+    public abstract Collection<RequestAndCompletionHandler> generateRequests();
+
+    public boolean hasUnsentRequests() {
+        return unsentRequests.iterator().hasNext();
+    }
+
+    @Override
+    public void shutdown() throws InterruptedException {
+        initiateShutdown();
+        networkClient.initiateClose();
+        awaitShutdown();
+        try {
+            networkClient.close();
+        } catch (IOException e) {
```

I was a bit surprised by this one. What's the reasoning here? I was wondering if we should just use `closeQuietly`. What do you think?

---

Nice, I didn't know about `closeQuietly` - I'll just integrate it here.
As for the reasoning - I wasn't sure whether it's OK to suppress the exception (as in theory it's not currently guaranteed that it strictly can't be thrown) so I decided to be overly conservative.